core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -669,13 +668,16 @@ public PlanOptimizers(
                         ruleStats,
                         statsCalculator,
                         estimatedExchangesCostCalculator,
-                        ImmutableSet.of(
-                                new RemoveEmptyUnionBranches(),
-                                new EvaluateEmptyIntersect(),
-                                new RemoveEmptyExceptBranches(),
-                                new RemoveRedundantIdentityProjections(),
-                                new PushAggregationThroughOuterJoin(),
-                                new ReplaceRedundantJoinWithSource())), // Run this after PredicatePushDown optimizer as it inlines filter constants
+                        ImmutableSet.<Rule<?>>builder()
+                                .add(new RemoveEmptyUnionBranches())
+                                .add(new EvaluateEmptyIntersect())
+                                .add(new RemoveEmptyExceptBranches())
+                                .add(new RemoveRedundantIdentityProjections())
+                                .add(new PushAggregationThroughOuterJoin())
+                                .add(new ReplaceRedundantJoinWithSource()) // Run this after PredicatePushDown optimizer as it inlines filter constants
+                                .add(new MultipleDistinctAggregationToMarkDistinct()) // Run this after aggregation pushdown
+                                .addAll(limitPushdownRules)
```

Comment:
```suggestion
                                .addAll(limitPushdownRules) // including through MarkDistinct
```

Reply:
Actually having this here results in test failures. @kasiafi looked deeply into that and the problem comes from extra call of `PushLimitThroughOuterJoin` and the fact that `applyJoin` will not consume same limit twice.
So without this optimization went

```
... Limit -> OuterJoin -> Scan(no limit)
(PushLimitThroughOuterJoin)
... Limit -> OuterJoin -> Limit -> Scan(no limit)
(PushLimitToTableScan)
... Limit -> OuterJoin -> Scan(with limit)

now join is pushable to TS
```

With this extra line the optimization continues from when it completed previously
```
... Limit -> OuterJoin -> Scan(with limit)
(PushLimitThroughOuterJoin)
... Limit -> OuterJoin -> Limit -> Scan(with limit)
(PushLimitToTableScan; applyJoin return Optional.empty() because of https://github.com/trinodb/trino/blob/408476e5d4067db61324f773510c27731373b91b/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java#L459-L461)
... Limit -> OuterJoin -> Limit -> Scan(with limit)

!!!!! join no longer pushable to table scan
```

This could be addressed with https://github.com/trinodb/trino/pull/9056 but I am not super convinced this is the way to go.

Reply:
@findepi, @kasiafi I suggest we go without limit pushdown here for now. It should not matter. What was to be pushed down should be already.
Do you feel it is good to be merged?

Reply:
The change in `PlanOptimizers` looks OK.